File: android/app/src/main/java/com/agcforge/videodownloader/data/model/Application.kt
package com.agcforge.videodownloader.data.model

import android.os.Parcelable
import com.google.gson.annotations.SerializedName
import kotlinx.parcelize.Parcelize

@Parcelize
data class Application(
    @SerializedName("id") val id: String,
    @SerializedName("name") val name: String,
    @SerializedName("package_name") val packageName: String,
    @SerializedName("version") val version: String? = null,
    @SerializedName("platform") val platform: String = "android",
    @SerializedName("enable_monetization") val enableMonetization: Boolean,
    @SerializedName("enable_admob") val enableAdmob: Boolean,
    @SerializedName("enable_unity_ad") val enableUnityAd: Boolean,
    @SerializedName("enable_start_app") val enableStartApp: Boolean,
    @SerializedName("enable_in_app_purchase") val enableInAppPurchase: Boolean,
    @SerializedName("admob_ad_unit_id") val admobAdUnitId: String? = null,
    @SerializedName("unity_ad_unit_id") val unityAdUnitId: String? = null,
    @SerializedName("start_app_ad_unit_id") val startAppAdUnitId: String? = null,
    @SerializedName("admob_banner_ad_unit_id") val admobBannerAdUnitId: String? = null,
    @SerializedName("admob_interstitial_ad_unit_id") val admobInterstitialAdUnitId: String? = null,
    @SerializedName("admob_native_ad_unit_id") val admobNativeAdUnitId: String? = null,
    @SerializedName("admob_rewarded_ad_unit_id") val admobRewardedAdUnitId: String? = null,
    @SerializedName("unity_banner_ad_unit_id") val unityBannerAdUnitId: String? = null,
    @SerializedName("unity_interstitial_ad_unit_id") val unityInterstitialAdUnitId: String? = null,
    @SerializedName("unity_native_ad_unit_id") val unityNativeAdUnitId: String? = null,
    @SerializedName("unity_rewarded_ad_unit_id") val unityRewardedAdUnitId: String? = null,
    @SerializedName("one_signal_id") val oneSignalId: String? = null,
    @SerializedName("is_active") val isActive: Boolean,
    @SerializedName("created_at") val createdAt: String,
    @SerializedName("updated_at") val updatedAt: String,
    @SerializedName("in_app_products") val inAppProducts: List<InAppProduct>? = null
) : Parcelable {

    fun isAdmobEnabled(): Boolean = enableMonetization && enableAdmob

    fun isUnityAdEnabled(): Boolean = enableMonetization && enableUnityAd

    fun isStartAppEnabled(): Boolean = enableMonetization && enableStartApp

    fun isInAppPurchaseEnabled(): Boolean = enableMonetization && enableInAppPurchase

    fun getAdmobBannerId(): String? = if (isAdmobEnabled()) admobBannerAdUnitId else null

    fun getAdmobInterstitialId(): String? = if (isAdmobEnabled()) admobInterstitialAdUnitId else null

    fun getAdmobRewardedId(): String? = if (isAdmobEnabled()) admobRewardedAdUnitId else null

    fun getUnityBannerId(): String? = if (isUnityAdEnabled()) unityBannerAdUnitId else null

    fun getUnityInterstitialId(): String? = if (isUnityAdEnabled()) unityInterstitialAdUnitId else null

    fun getUnityRewardedId(): String? = if (isUnityAdEnabled()) unityRewardedAdUnitId else null

    fun isValid(): Boolean = id.isNotEmpty() && name.isNotEmpty() && packageName.isNotEmpty()
}

@Parcelize
data class InAppProduct(
    @SerializedName("id") val id: String,
    @SerializedName("app_id") val appId: String,
    @SerializedName("product_id") val productId: String? = null,
    @SerializedName("product_type") val productType: String? = null,
    @SerializedName("sku_code") val skuCode: String? = null,
    @SerializedName("title") val title: String? = null,
    @SerializedName("description") val description: String? = null,
    @SerializedName("price") val price: Double? = null,
    @SerializedName("currency") val currency: String? = null,
    @SerializedName("billing_period") val billingPeriod: String? = null,
    @SerializedName("trial_period_days") val trialPeriodDays: Int? = null,
    @SerializedName("is_active") val isActive: Boolean,
    @SerializedName("is_featured") val isFeatured: Boolean,
    @SerializedName("sort_order") val sortOrder: Int,
    @SerializedName("created_at") val createdAt: String,
    @SerializedName("updated_at") val updatedAt: String
) : Parcelable {

    fun isValid(): Boolean =
        id.isNotEmpty() &&
                appId.isNotEmpty() &&
                productId?.isNotEmpty() == true &&
                productType?.isNotEmpty() == true &&
                skuCode?.isNotEmpty() == true &&
                title?.isNotEmpty() == true &&
                description?.isNotEmpty() == true &&
                price != null &&
                currency?.isNotEmpty() == true &&
                billingPeriod?.isNotEmpty() == true
}

File: backend/internal/model/application.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type Application struct {
	ID                        uuid.UUID `json:"id" db:"id"`
	Name                      string    `json:"name" db:"name"`
	PackageName               string    `json:"package_name" db:"package_name"`
	APIKey                    string    `json:"api_key" db:"api_key"`
	SecretKey                 string    `json:"secret_key" db:"secret_key"`
	Version                   *string   `json:"version" db:"version"`   // Nullable
	Platform                  string    `json:"platform" db:"platform"` // default: android
	EnableMonetization        bool      `json:"enable_monetization" db:"enable_monetization"`
	EnableAdmob               bool      `json:"enable_admob" db:"enable_admob"`
	EnableUnityAd             bool      `json:"enable_unity_ad" db:"enable_unity_ad"`
	EnableStartApp            bool      `json:"enable_start_app" db:"enable_start_app"`
	EnableInAppPurchase       bool      `json:"enable_in_app_purchase" db:"enable_in_app_purchase"`
	AdmobAdUnitID             *string   `json:"admob_ad_unit_id" db:"admob_ad_unit_id"`
	UnityAdUnitID             *string   `json:"unity_ad_unit_id" db:"unity_ad_unit_id"`
	StartAppAdUnitID          *string   `json:"start_app_ad_unit_id" db:"start_app_ad_unit_id"`
	AdmobBannerAdUnitID       *string   `json:"admob_banner_ad_unit_id" db:"admob_banner_ad_unit_id"`
	AdmobInterstitialAdUnitID *string   `json:"admob_interstitial_ad_unit_id" db:"admob_interstitial_ad_unit_id"`
	AdmobNativeAdUnitID       *string   `json:"admob_native_ad_unit_id" db:"admob_native_ad_unit_id"`
	AdmobRewardedAdUnitID     *string   `json:"admob_rewarded_ad_unit_id" db:"admob_rewarded_ad_unit_id"`
	UnityBannerAdUnitID       *string   `json:"unity_banner_ad_unit_id" db:"unity_banner_ad_unit_id"`
	UnityInterstitialAdUnitID *string   `json:"unity_interstitial_ad_unit_id" db:"unity_interstitial_ad_unit_id"`
	UnityNativeAdUnitID       *string   `json:"unity_native_ad_unit_id" db:"unity_native_ad_unit_id"`
	UnityRewardedAdUnitID     *string   `json:"unity_rewarded_ad_unit_id" db:"unity_rewarded_ad_unit_id"`
	OneSignalID               *string   `json:"one_signal_id" db:"one_signal_id"`
	IsActive                  bool      `json:"is_active" db:"is_active"`
	CreatedAt                 time.Time `json:"created_at" db:"created_at"`
	UpdatedAt                 time.Time `json:"updated_at" db:"updated_at"`

	// Relations
	InAppProducts []InAppProduct `json:"in_app_products,omitempty" db:"-"`
	Subscriptions []Subscription `json:"subscriptions,omitempty" db:"-"`
	Downloads     []DownloadTask `json:"downloads,omitempty" db:"-"`
}

type InAppProduct struct {
	ID              uuid.UUID       `json:"id" db:"id"`
	AppID           uuid.UUID       `json:"app_id" db:"app_id"`
	ProductID       *string         `json:"product_id,omitempty" db:"product_id"`
	ProductType     *string         `json:"product_type,omitempty" db:"product_type"`
	SkuCode         *string         `json:"sku_code,omitempty" db:"sku_code"`
	Title           *string         `json:"title,omitempty" db:"title"`
	Description     *string         `json:"description,omitempty" db:"description"`
	Price           *float64        `json:"price,omitempty" db:"price"`
	Currency        *string         `json:"currency,omitempty" db:"currency"`
	BillingPeriod   *string         `json:"billing_period,omitempty" db:"billing_period"`
	TrialPeriodDays *int32          `json:"trial_period_days,omitempty" db:"trial_period_days"`
	IsActive        bool            `json:"is_active" db:"is_active"`
	IsFeatured      bool            `json:"is_featured" db:"is_featured"`
	SortOrder       int32           `json:"sort_order" db:"sort_order"`
	Features        *map[string]any `json:"features,omitempty" db:"features"`
	CreatedAt       time.Time       `json:"created_at" db:"created_at"`
	UpdatedAt       time.Time       `json:"updated_at" db:"updated_at"`
}

type Subscription struct {
	ID                    uuid.UUID  `json:"id" db:"id"`
	UserID                *uuid.UUID `json:"user_id,omitempty" db:"user_id"`
	AppID                 *uuid.UUID `json:"app_id,omitempty" db:"app_id"`
	OriginalTransactionID string     `json:"original_transaction_id" db:"original_transaction_id"`
	ProductID             string     `json:"product_id" db:"product_id"`
	PurchaseToken         string     `json:"purchase_token" db:"purchase_token"`
	Platform              string     `json:"platform" db:"platform"`
	StartTime             time.Time  `json:"start_time" db:"start_time"`
	EndTime               time.Time  `json:"end_time" db:"end_time"`
	Status                string     `json:"status" db:"status"`
	AutoRenew             bool       `json:"auto_renew" db:"auto_renew"`
	CreatedAt             time.Time  `json:"created_at" db:"created_at"`
	UpdatedAt             time.Time  `json:"updated_at" db:"updated_at"`

	// Relations
	User         *User         `json:"user,omitempty" db:"-"`
	Transactions []Transaction `json:"transactions,omitempty" db:"-"`
}

type Transaction struct {
	ID               uuid.UUID      `json:"id" db:"id"`
	UserID           *uuid.UUID     `json:"user_id,omitempty" db:"user_id"`
	AppID            *uuid.UUID     `json:"app_id,omitempty" db:"app_id"`
	SubscriptionID   *uuid.UUID     `json:"subscription_id,omitempty" db:"subscription_id"`
	Amount           float64        `json:"amount" db:"amount"`
	Currency         string         `json:"currency" db:"currency"`
	Provider         string         `json:"provider" db:"provider"`
	Status           string         `json:"status" db:"status"`
	ProviderResponse map[string]any `json:"provider_response" db:"provider_response"`
	CreatedAt        time.Time      `json:"created_at" db:"created_at"`

	// Relations
	User         *User         `json:"user,omitempty" db:"-"`
	Subscription *Subscription `json:"subscription,omitempty" db:"-"`
}

type RegisterAppResponse struct {
	Name        string `json:"name"`
	PackageName string `json:"package_name"`
	APIKey      string `json:"api_key"`
	SecretKey   string `json:"secret_key"`
}

type RegisterAppRequest struct {
	Name                      string  `json:"name" validate:"required"`
	PackageName               string  `json:"package_name" validate:"required"`
	Version                   string  `json:"version"`
	Platform                  string  `json:"platform"`
	EnableMonetization        bool    `json:"enable_monetization"`
	EnableAdmob               bool    `json:"enable_admob"`
	EnableUnityAd             bool    `json:"enable_unity_ad"`
	EnableStartApp            bool    `json:"enable_start_app"`
	EnableInAppPurchase       bool    `json:"enable_in_app_purchase"`
	AdmobAdUnitID             *string `json:"admob_ad_unit_id"`
	UnityAdUnitID             *string `json:"unity_ad_unit_id"`
	StartAppAdUnitID          *string `json:"start_app_ad_unit_id"`
	AdmobBannerAdUnitID       *string `json:"admob_banner_ad_unit_id"`
	AdmobInterstitialAdUnitID *string `json:"admob_interstitial_ad_unit_id"`
	AdmobNativeAdUnitID       *string `json:"admob_native_ad_unit_id"`
	AdmobRewardedAdUnitID     *string `json:"admob_rewarded_ad_unit_id"`
	UnityBannerAdUnitID       *string `json:"unity_banner_ad_unit_id"`
	UnityInterstitialAdUnitID *string `json:"unity_interstitial_ad_unit_id"`
	UnityNativeAdUnitID       *string `json:"unity_native_ad_unit_id"`
	UnityRewardedAdUnitID     *string `json:"unity_rewarded_ad_unit_id"`
	OneSignalID               *string `json:"one_signal_id"`
	IsActive                  bool    `json:"is_active"`
}

type BulkDeleteAppsRequest struct {
	IDs []uuid.UUID `json:"ids" validate:"required,min=1"`
}
